vmm/src/lib.rs

```diff
@@ -1081,6 +1229,132 @@ impl Vmm {
+        let mut existing_memory_files = None;
+
+        loop {
+            let req = Request::read_from(&mut socket)?;
+            match req.command() {
+                Command::Invalid => info!("Invalid Command Received"),
+                Command::Start => {
+                    info!("Start Command Received");
+                    started = true;
+
+                    Response::ok().write_to(&mut socket)?;
+                }
+                Command::Config => {
+                    info!("Config Command Received");
+
+                    if !started {
+                        warn!("Migration not started yet");
+                        Response::error().write_to(&mut socket)?;
+                        continue;
+                    }
+                    memory_manager = Some(self.vm_receive_config(
+                        &req,
+                        &mut socket,
+                        existing_memory_files.take(),
+                    )?);
+                }
+                Command::State => {
+                    info!("State Command Received");
+
+                    if !started {
+                        warn!("Migration not started yet");
+                        Response::error().write_to(&mut socket)?;
+                        continue;
+                    }
+                    if let Some(mm) = memory_manager.take() {
+                        self.vm_receive_state(&req, &mut socket, mm)?;
+                    } else {
+                        warn!("Configuration not sent yet");
+                        Response::error().write_to(&mut socket)?;
+                    }
+                }
+                Command::Memory => {
+                    info!("Memory Command Received");
+
+                    if !started {
+                        warn!("Migration not started yet");
+                        Response::error().write_to(&mut socket)?;
+                        continue;
+                    }
+                    if let Some(mm) = memory_manager.as_ref() {
+                        self.vm_receive_memory(&req, &mut socket, &mut mm.lock().unwrap())?;
+                    } else {
+                        warn!("Configuration not sent yet");
+                        Response::error().write_to(&mut socket)?;
+                    }
+                }
+                Command::MemoryFd => {
+                    info!("MemoryFd Command Received");
+
+                    if !started {
+                        warn!("Migration not started yet");
+                        Response::error().write_to(&mut socket).map_err(|e| {
+                            MigratableError::MigrateReceive(anyhow!(
+                                "Error sending response: {}",
+                                e
+                            ))
+                        })?;
+                        continue;
+                    }
+
+                    if let SocketStream::Unix(unix_socket) = &mut socket {
```

Comment:
> The `MemoryFd` command is used with local migration only, and hence won't work with the tcp socket. Let's make sure we report an error and abort the migration for that case.

Yes, the MemoryFd command is only used for local migration. There is no need for any tcp processing here, because tcp live migration will never call the MemoryFd command.

Reply:
Maybe I was not clear about the changes that I suggested earlier. It is better to use `match` instead of `if let` here, so that you can explicitly report errors when TCP socket is used with local migration (like what you have done from `send_migration`).

Reply:
> Maybe I was not clear about the changes that I suggested earlier. It is better to use `match` instead of `if let` here, so that you can explicitly report errors when TCP socket is used with local migration (like what you have done from `send_migration`).

I don't quite understand this description. I can change the `if let` to `match`, but there will not be any handling of TCP here, because the handling of TCP here will be redundant. If TCP sockets are used with local migration, an error will be reported in `if send_data_migration.local`.

```
        if send_data_migration.local {
            match &mut socket {
                SocketStream::Unix(unix_socket) => {
                    // Proceed with sending memory file descriptors over UNIX socket
                    vm.send_memory_fds(unix_socket)?;
                }
                SocketStream::Tcp(_tcp_socket) => {
                    return Err(MigratableError::MigrateSend(anyhow!(
                        "--local option is not supported with TCP sockets",
                    )));
                }
            }
        }
```